File: 801-is-graph-bipartite/is-graph-bipartite.cpp
class Solution {
public:
    bool fun(vector<vector<int>>& graph,vector<int>& visited,int i,int color)
    {
        visited[i]=color;
        
        for(auto& j : graph[i])
        {
            if(visited[j]==visited[i])
            {
                return false;
            }
            if(visited[j]==-1)
            {
                if(fun(graph,visited,j,1-color)==false){
                return false;}
            }
        }
        return true;
    }
    bool isBipartite(vector<vector<int>>& graph) {
        int n=graph.size();
        vector<int> visited(n,-1);
        for(int i=0;i<n;i++)
        {
            if(visited[i]==-1)
            {
                if(!fun(graph,visited,i,1))
                {
                     return false;
                }   
            }
        }
        return true;
    }
};

File: 801-is-graph-bipartite/is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        int[] color = new int[n];
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (color[i] == 0) {
                color[i] = 1;
                q.offer(i);
                while (!q.isEmpty()) {
                    int p = q.poll();
                    for (int neighbor : graph[p]) {
                        if (color[neighbor] == 0) {

                            color[neighbor] = -color[p];
                            q.offer(neighbor);
                        } else if (color[neighbor] == color[p]) {

                            return false;
                        }
                    }
                }
            }

        }
        return true;
    }
}